pkg/git/strategy/proxy/strategy_proxy_test.go

```diff
@@ -0,0 +1,282 @@
+						*proxyGotRequest = true
+						req.Host = u.Host
+						req.URL.Host = req.Host
+						return req, nil
+					}
+					//reject if it isnt our request
+					return req, goproxy.NewResponse(req, goproxy.ContentTypeText, http.StatusForbidden, "")
+				}
+				proxy.OnRequest().Do(proxyHandler)
+
+				return &git.AuthOptions{
+						Transport: git.HTTP,
+						Username:  username,
+						Password:  password,
+					}, func() {
+						os.RemoveAll(gitServer.Root())
+						gitServer.StopHTTP()
+					}
+			},
+			shortTimeout:  false,
+			wantUsedProxy: true,
+			wantError:     false,
+		},
+		{
+			name:    "gogit_HTTPS_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "https://github.com/git-fixtures/basic",
+			branch:  "master",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, cleanupFunc) {
+				var proxyHandler goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
+					userAgent := ctx.Req.Header.Get("User-Agent")
+					if strings.Contains(host, "github.com") && strings.Contains(userAgent, "Go-http-client") {
+						*proxyGotRequest = true
+						return goproxy.OkConnect, host
+					}
+					//reject if it isnt our request
+					return goproxy.RejectConnect, host
+				}
+				proxy.OnRequest().HandleConnect(proxyHandler)
+
+				// go-git does not allow to use an HTTPS proxy and a custom root CA at the same time.
+				// See https://github.com/fluxcd/source-controller/pull/524#issuecomment-1006673163.
+				return nil, func() {}
+			},
+			shortTimeout:  false,
+			wantUsedProxy: true,
+			wantError:     false,
+		},
+		{
+			name:    "gogit_NO_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "https://192.0.2.1/bar/test-reponame",
+			branch:  "main",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, cleanupFunc) {
+				var proxyHandler goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
+					// we shouldnt hit the proxy so we just want to check for any interaction, then reject
+					*proxyGotRequest = true
+					return goproxy.RejectConnect, host
+				}
+				proxy.OnRequest().HandleConnect(proxyHandler)
+
+				return nil, func() {}
+			},
+			shortTimeout:  true,
+			wantUsedProxy: false,
+			wantError:     true,
+		},
+		// TODO: Add a NO_PROXY test for libgit2 once the version of libgit2 used by the source controller is updated to a version that includes
+		// the NO_PROXY functionlity
+	}
+
+	testFunc := func(tt testCase) func(t *testing.T) {
+		return func(t *testing.T) {
+			g := NewWithT(t)
+
+			// Run a proxy server.
+			proxy := goproxy.NewProxyHttpServer()
+			proxy.Verbose = true
+
+			proxyGotRequest := false
+			authOpts, cleanup := tt.setupGitProxy(g, proxy, &proxyGotRequest)
+			defer cleanup()
+
+			proxyServer := http.Server{
+				Addr:    proxyAddr,
```

Comment:
Instead of setting a fixed port for proxy address, which could be in use in some development machines, let's make this dynamic and select a free port.
Since the proxy settings are cached, we'll have to ensure that the same port is used across all the test cases. So, let's get a free port after the test cases code and before running the tests, something like:

```go
	// Get a free port for proxy.
	g := NewWithT(t)
	l, err := net.Listen("tcp", ":0")
	g.Expect(err).ToNot(HaveOccurred())
	proxyAddr := fmt.Sprintf("localhost:%d", l.Addr().(*net.TCPAddr).Port)
	g.Expect(l.Close()).ToNot(HaveOccurred())
```
Closing the listener here so that we can listen and close per test case and use it with different proxy servers.

With this, the `proxyAddr := "localhost:9999"` at the top of the test function can be removed.

As a side-effect of this, running tests with `-count` value more than 1 would just fail. But that should be clear by the comment at the top of the function which states that the proxy configurations are cached by go.